src/SleighInstruction.cpp

```diff
@@ -0,0 +1,462 @@
+			curflags = FlowFlags(curflags & (~(FLOW_CROSSBUILD | FLOW_LABEL | FLOW_NO_FALLTHRU)));
+			curflags = FlowFlags(curflags | rec->flowFlags);
+		}
+	}
+	return curflags;
+}
+
+void SleighInstruction::gatherFlows(std::vector<Address> &res, ParserContext *parsecontext,
+                                    int secnum)
+{
+	std::vector<FlowRecord *> curlist;
+	if(secnum < 0)
+		curlist = flowStateList;
+	else if((!flowStateListNamed.empty()) && (secnum < flowStateListNamed.size()))
+		curlist = flowStateListNamed[secnum];
+
+	if(curlist.empty())
+		return;
+
+	for(FlowRecord *rec: curlist)
+	{
+		if((rec->flowFlags & FLOW_CROSSBUILD) != 0)
+		{
+			SubParserWalker walker(parsecontext);
+			walker.subTreeState(rec->addressnode);
+
+			VarnodeTpl *vn = rec->op->getIn(0);
+			AddrSpace *spc = vn->getSpace().fixSpace(walker);
+			uintb addr = spc->wrapOffset(vn->getOffset().fix(walker));
+			Address newaddr(spc, addr);
+			int newsecnum = rec->op->getIn(1)->getOffset().getReal();
+			SleighInstruction *crossproto = sleigh->getInstruction(newaddr);
+			crossproto->gatherFlows(res, crossproto->protoContext, newsecnum);
+		}
+		else if((rec->flowFlags & (FLOW_JUMPOUT | FLOW_CALL)) != 0)
+		{
+			FixedHandle &hand = rec->addressnode->hand;
+			if(!handleIsInvalid(hand) && hand.offset_space == nullptr)
+			{
+				Address addr = getHandleAddr(hand, parsecontext->getAddr().getSpace());
+				res.push_back(addr);
+			}
+		}
+	}
+}
+
+Address SleighInstruction::getHandleAddr(FixedHandle &hand, AddrSpace *curSpace)
+{
+	if(handleIsInvalid(hand) || hand.space->getType() == spacetype::IPTR_INTERNAL ||
+	   hand.offset_space != nullptr)
+		return Address();
+
+	Address newaddr(hand.space, hand.space->wrapOffset(hand.offset_offset));
+
+	newaddr.toPhysical();
+
+	// if we are in an address space, translate it
+	// if (curSpace.isOverlaySpace()) {
+	// newaddr = curSpace.getOverlayAddress(newaddr);
+	// }
+	return newaddr;
+}
+
+bool SleighInstruction::handleIsInvalid(FixedHandle &hand)
+{
+	return hand.space == nullptr;
+}
+
+FlowType SleighInstruction::getFlowType()
+{
+	if(!hasCrossBuilds)
+		return flowType;
+
+	return convertFlowFlags(gatherFlags(FlowFlags(0), -1));
+}
+
+std::vector<Address> SleighInstruction::getFlows()
+{
+	std::vector<Address> addresses;
+	if(flowStateList.empty())
+		return addresses;
+
+	SleighParserContext *pos = protoContext;
+	pos->applyCommits();
+	pos->clearCommits();
+	gatherFlows(addresses, pos, -1);
+
+	return addresses;
+}
+
+Address SleighInstruction::getFallThrough()
+{
+	if(flowTypeHasFallthrough(flowType))
+		return baseaddr + getFallThroughOffset();
+
+	return Address();
+}
+
+int SleighInstruction::getFallThroughOffset()
+{
+	if(delaySlotByteCnt <= 0)
+		return getLength();
+
+	int offset = getLength();
+	int bytecount = 0;
+	do
+	{
+		Address off_addr = baseaddr + offset;
+		SleighInstruction ins(sleigh, off_addr);
+		int len = ins.getLength();
+		offset += len;
+		bytecount += len;
+	} while(bytecount < delaySlotByteCnt);
```

Comment:
Probably a check with len < 0 should be added?

Reply:
Agree